### ISSUE TITLE
Train a Toolbox model on (noisy) UIE imagery and contrast efficacy with Toolbox model trained on hand-edited imagery 

### ISSUE THREAD

**Issue by zhrandell** [Seattle-Aquarium/CCR_image_processing#11]
Now that we have a image processing algorithm that performs well (see [here](https://github.com/keenanjohnson/underwater-auto-image-encoder), https://github.com/Seattle-Aquarium/CCR_image_processing/issues/8, and https://github.com/Seattle-Aquarium/CCR_image_processing/issues/10) let's conduct some preliminary tests to see how Toolbox does with it. 

Specifically, we have a [CoralNet-Toolbox](https://github.com/Jordan-Pierce/CoralNet-Toolbox) classification model (YOLOv11s-cls) trained on 25 distinct categories of algae and substrate, with ~500 manual annotations for each category. The model is performing well, with 96% accuracy when applied to the "test" data during model training (the confusion matrix and model weights can be found [here](https://www.dropbox.com/scl/fo/c46wxcqwqsh8tqo4b5w9y/AEZvkwmiLmnzv49PJY0hx6A?rlkey=aohp9lcm6feskzkckvbb8atfn&dl=0), and you can see some of the image patches from the training dataset [here](https://www.dropbox.com/scl/fo/gs8eghjm2bfsufjk31cws/ACxvQdKkZhhn6jrjdzJqPto?rlkey=u8s7kqjqyk1fgrs4x9tvlgukz&dl=0)).

However, this model was trained on imagery hand-edited in Adobe Lightroom, including usage of the Denoise Feature, which reduces noise while maintaining sharpness and boundaries, as discussed in https://github.com/Seattle-Aquarium/CCR_image_processing/issues/8. 

Our objective here is to evaluate the efficacy of the existing trained Toolbox model when applied to n=20 images that have been (**1**) edited by-hand in Adobe Lightroom vs (**2**) processed by the UIE. Specifically, we'll need to: 

- [x] upload both images sets to Toolbox and create identical grids of 224 x 224 patches (likely n=50 uniformly distributed patches per image)
- [x] apply the Toolbox model to our n=20 hand-edited images
- [x] apply the Toolbox model to our n=20 UIE edited images
- [ ] manually review model predictions for both image sets and download resulting .csv files
- [ ] analyze differences in model % accuracy across categories when comparing hand-edited vs UIE predictions
- [ ] the above ^ will require revising a R script to properly analyze the number of correct vs incorrect Toolbox model predictions per category

**Comment by zhrandell** [comment]
excellent, thanks Megan. It sounds like the only piece we need new code for is to isolate the .GPRs that correspond to the file names in `dataset.csv` and copy them over to a new file (easy!). Copying over much of what you said, I think the total workflow is:

- [ ] copy .GPRs present within [dataset.csv](https://www.dropbox.com/scl/fi/i0xjc7cuaz9d7atqak3jx/dataset.csv?rlkey=8mpo06iazdp2zp5m41ic4fbm8&dl=0) into new folder (you'll need to ignore the `.jpg` extension)
- [ ] run those GPRs through the UIE pipeline
- [ ] upload UIE processed imagery into Toolbox
- [ ] create a new training dataset using existing `Row` & `Column` deliniators of specific, already labeled patches
- [ ] run a Toolbox model training with this new dataset
- [ ] compare new model training efficacy to existing model trained on Adobe Lightroom imagery. 

I'm going to tag @reidtosa in on the first two bullet points - this should be super simple to do - could you please isolate those .GPRs and run them through the UIE pipeline next week when you're back in the office?

**Comment by m-h-williams** [comment]
I trained two YOLO classification models for comparison: one using UIE images without denoise and one using hand-edited images denoised in Lightroom.

### **Model 1: UIE model**

- **Dataset:** [UIE_dataset_rebalanced](https://www.dropbox.com/scl/fo/zhnppemc8ljz2amoywu6m/AELTLtY_i6p9bhaJPZjbKRQ?rlkey=6stllsmmve1jqd504cz61npc1&st=th4x8p1q&dl=0)
   - Source: [UIE images](https://www.dropbox.com/scl/fo/pxkd3mta0vyl2qalnpxih/AH_eWFfykOEJv4CwntF8D9s?rlkey=c7cqbnauj2dvhnzwo7a27a5ct&st=qt5232ob&dl=0) (n = 1099 images), 
   - Imported:  [labelset.json](https://www.dropbox.com/scl/fi/2hw2h3zt8njdwts82jdky/labelset_current.json?rlkey=8wwm5xzdjsvi55f0o2gh3rmdj&st=zn0fekf6&dl=0), and the `dataset.csv` annotation file from the  [EB_dataset_pt2_rebalanced](https://www.dropbox.com/scl/fo/ku2usjmzlo58krdq7240p/ACroAaT_aIpsk5V_YK9U5vM?rlkey=tumdbg37r7zg2bzzy4fqr2j8t&st=wtqn97gy&dl=0) classify dataset. 
   - Balanced: max of 500 training annotations,  0.7/ 0.2/  0.1 (train/val/test) using [rebalance_dataset.py](https://www.dropbox.com/scl/fi/4aw4bom7p1qo6q7ngvajn/rebalance_dataset.py?rlkey=pfkhh0qdvlso6hp5lpmq8hsbx&st=gw9dwhz2&dl=0) 
   - Resulting label distribution in dataset is [here](https://www.dropbox.com/scl/fi/34bbab4pvtmw80c67124b/label_distribution.csv?rlkey=0zh0uge9ihrjygn5ihy51ohfk&st=1zdx913a&dl=0)
- Training setup
   - Base YOLO model: `YOLOs-cls` 
   - Epochs: 100
   - [Augmentations](https://docs.ultralytics.com/modes/train/#train-settings): `hsv_h = 0.015`, `hsv_s = 0.7`, `hsv_v = 0.4`, `flipud = 0.5` 
- **Results:** 
   - Model: [UIE_rebalanced_model](https://www.dropbox.com/scl/fo/nkpig7kwvblaksogmmr3t/AE6p1Yc6oKJM-NHeQ98zwb0?rlkey=3tzyeu4lc7ry2lest498eyah7&st=y8sglq8b&dl=0)
   - Top-1 test accuracy = 0.817
   - Top-5 test accuracy = 0.96

### **Model 2: Hand-edited model**

- **Dataset:** [hand_edited_dataset_rebalanced](https://www.dropbox.com/scl/fo/e5a9rz55fkmxgug32t2w8/AMQ5QrA6Mcf4Q2tz-nrMxK4?rlkey=p179q9et0cy1pw66ljwdnlrl6&st=oo7spzvx&dl=0)
   - Source: [hand-edited images ](https://www.dropbox.com/scl/fo/lyeh433n8ovqzjqrflsnq/ALi0zOiTvSTyAy0ozM3_dlY?rlkey=0ccblk42igoxdvmmjky2xpyw8&st=lge7kg6n&dl=0) with lightroom denoise, matched to UIE image names (n = 1099 images)  
   - Imported: same `labelset.json` and `dataset.csv` annotation file from the `EB_dataset_pt2_rebalanced` classify dataset. 
   - Balanced: max of 500 training annotations,  0.7/ 0.2/  0.1 (train/val/test) 
   - Resulting label distribution [here](https://www.dropbox.com/scl/fi/62fp63zo8yo7dqm7htt6u/label_distribution.csv?rlkey=n44azgz1bvtyxxo5gqltm76tk&st=rm2uhxf9&dl=0) 
- **Training setup**:
   - Base YOLO model: `YOLOs-cls`
   - Epochs: 100
   - Augmentations: `hsv_h = 0.015`, `hsv_s = 0.7`, `hsv_v = 0.4`, `flipud = 0.5` 
- **Results:**
   -  Model: [hand_edited_rebalanced_model](https://www.dropbox.com/scl/fo/603rr2zu1z887k3frt5qm/AIJzKgdD6hqKZY2apqOLbu0?rlkey=3v4jtce8rqxecbz9glhe2bjhb&st=1595ty0m&dl=0)
   - Top-1 test accuracy = 0.932
   - Top-5 test accuracy = 0.994

### Findings
- Hand-edited images (denoised in Lightroom) achieved >10% improvement in Top-1 accuracy compared to UIE images.

**Note**: Additional details for each model (e.g., confusion matrix, per-epoch training results CSV) are available in the model folders linked above.

**Comment by keenanjohnson** [comment]
@m-h-williams thanks so much! It's really helpful to have a metric like this we can work on improving rather than relying strictly on visual review. Is it complex or impractical for me to learn how train one of these models so that I can help work on training a model that closes the gap to the human edited images without creating lots of work for you all haha?

**Comment by keenanjohnson** [comment]
Thank you! I'll try that out !

**Comment by keenanjohnson** [comment]
Yes agree that the initial effort has been completed! Given that we have only evaluated one of the models from the initial literature study that Tonio created, it's possible that a different technique might perform better if trained. However, the other variable seems to be the noise and sharpness, which are a different problem than the lighting enhancement covered by these papers. I may still evaluate one or two other techniques to see if they improve the accuracy and I believe that @zhrandell mentioned @reidtosa was going to try some tests with the sharpening to see if that improves the accuracy metric which will help us determine if we need to explore different techniques for color adjustment or spend more effort on sharpening / noise removal.
